aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -1,72 +1,278 @@
-  // See [Note: hacky wrapper removal for optional tensor]
-  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
-  const Tensor& weight = *weight_maybe_owned;
-  const Tensor& bias = c10::value_or_else(bias_opt, [] {return Tensor();});
-  const Tensor& running_mean = c10::value_or_else(running_mean_opt, [] {return Tensor();});
-  const Tensor& running_var = c10::value_or_else(running_var_opt, [] {return Tensor();});
+  const bool has_running_mean = (running_mean_opt.has_value() && running_mean_opt->defined());
+  const bool has_running_var = (running_mean_opt.has_value() && running_mean_opt->defined());
+  TORCH_CHECK(has_running_mean == has_running_var);
 
-  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, self.scalar_type(), "batch_norm_cuda", [&] {
-    auto mean_st = running_mean.dtype();
-    auto var_st = running_var.dtype();
-    TORCH_CHECK(mean_st == var_st, "running_mean and running_var need to have the same data types");
-    bool is_half_float = std::is_same<scalar_t, at::Half>::value && mean_st == at::kFloat;
-    bool is_bfloat16_float = std::is_same<scalar_t, at::BFloat16>::value && mean_st == at::kFloat;
-    if (cuda::detail::canUse32BitIndexMath(self)) {
-      if (is_half_float || is_bfloat16_float) {
-        batch_norm_cuda_template<scalar_t, float, int32_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
-      } else {
-        batch_norm_cuda_template<scalar_t, scalar_t, int32_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
-      }
+  if (train) {
+    batch_norm_mean_var(self, save_mean, save_invstd);
+    if (has_running_mean) {
+      const int64_t N = self.numel() / save_mean.numel();
+      batch_norm_update_stats_and_invert(
+          save_mean, save_invstd, *running_mean_opt, *running_var_opt,
+          momentum, epsilon, N);
     } else {
-      if (is_half_float || is_bfloat16_float) {
-        batch_norm_cuda_template<scalar_t, float, int64_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
-      } else {
-        batch_norm_cuda_template<scalar_t, scalar_t, int64_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
-      }
+      batch_norm_calc_invstd(save_invstd, save_invstd, epsilon);
     }
-  });
+  } else {
+    TORCH_CHECK(has_running_mean);
+    at::native::resize_output(save_mean, running_mean_opt->sizes());
+    save_mean.copy_(*running_mean_opt);
+    batch_norm_calc_invstd(save_invstd, running_var_opt.value(), epsilon);
```

Comment:
I assume the number of channels is small relative to `self.numel()`, so this introduces fairly minimal overhead in general.

The old code calculates invstd inside of the element-wise kernel which means far more redundant divide and sqrt operations are done (`O(self.numel())` instead of `O(self.size(1))`). For the example in the issue, `nvprof` shows separate kernels as being marginally _faster_. The copy and invstd kernels take on average 1 us each, but the `transform_input_kernel` is ~8 us faster.

Another reason to do it this way is avoid compiling two different versions of the element-wise kernel.

Reply:
Yeah, those are all good reasons, I agree. It's not so much kernel time itself, as launch latency that comes with each kernel (on the order of 5-7 us) and `copy_` operation itself (another couple microseconds, give or take) and that will show up if code is cpu bound. That said, I agree that perf difference is pretty minimal, and currently code is more modular, so let's leave it this way.